nntrainer/layers/multi_head_attention_layer.cpp

```diff
@@ -73,10 +73,317 @@ enum AttentionParams {
+    weight_regularizer_constant, weight_decay, "fc_weight", true);
+  if (!disable_bias) {
+    TensorDim fc_bias_dim({1, 1, 1, output_shape});
+    weight_idx[AttentionParams::fc_bias] = context.requestWeight(
+      fc_bias_dim, weight_initializer, weight_regularizer,
+      weight_regularizer_constant, weight_decay, "fc_bias", true);
+  }
+
+  TensorDim projected_query_dim(
+    {batch_size, 1, query_height, num_heads * projected_query_dim_prop});
+  weight_idx[AttentionParams::projected_query] = context.requestTensor(
+    projected_query_dim, "projected_query", Tensor::Initializer::NONE, false,
+    TensorLifespan::FORWARD_FUNC_LIFESPAN);
+  weight_idx[AttentionParams::d_projected_query] = context.requestTensor(
+    projected_query_dim, "d_projected_query", Tensor::Initializer::NONE, false,
+    TensorLifespan::BACKWARD_FUNC_LIFESPAN);
+  TensorDim projected_key_dim(
+    {batch_size, 1, key_height, num_heads * projected_key_dim_prop});
+  weight_idx[AttentionParams::projected_key] = context.requestTensor(
+    projected_key_dim, "projected_key", Tensor::Initializer::NONE, true,
+    TensorLifespan::FORWARD_FUNC_LIFESPAN);
+  weight_idx[AttentionParams::d_projected_key] = context.requestTensor(
+    projected_key_dim, "d_projected_key", Tensor::Initializer::NONE, false,
+    TensorLifespan::BACKWARD_FUNC_LIFESPAN);
+  TensorDim projected_value_dim(
+    {batch_size, 1, value_height, num_heads * projected_value_dim_prop});
+  weight_idx[AttentionParams::projected_value] = context.requestTensor(
+    projected_value_dim, "projected_value", Tensor::Initializer::NONE, true,
+    TensorLifespan::FORWARD_FUNC_LIFESPAN);
+  weight_idx[AttentionParams::d_projected_value] = context.requestTensor(
+    projected_value_dim, "d_projected_value", Tensor::Initializer::NONE, false,
+    TensorLifespan::BACKWARD_FUNC_LIFESPAN);
+
+  TensorDim transposed_projected_query_dim(
+    {batch_size, num_heads, query_height, projected_query_dim_prop});
+  weight_idx[AttentionParams::transposed_projected_query] =
+    context.requestTensor(
+      transposed_projected_query_dim, "transposed_projected_query",
+      Tensor::Initializer::NONE, true, TensorLifespan::ITERATION_LIFESPAN);
+  TensorDim transposed_projected_key_dim(
+    {batch_size, num_heads, key_height, projected_key_dim_prop});
+  weight_idx[AttentionParams::transposed_projected_key] = context.requestTensor(
+    transposed_projected_key_dim, "transposed_projected_key",
+    Tensor::Initializer::NONE, true, TensorLifespan::ITERATION_LIFESPAN);
+  TensorDim transposed_projected_value_dim(
+    {batch_size, num_heads, value_height, projected_value_dim_prop});
+  weight_idx[AttentionParams::transposed_projected_value] =
+    context.requestTensor(
+      transposed_projected_value_dim, "transposed_projected_value",
+      Tensor::Initializer::NONE, true, TensorLifespan::ITERATION_LIFESPAN);
+
+  TensorDim attention_score_dim(
+    {batch_size, num_heads, query_height, key_height});
```

Comment:
please check if we could save the memory further.

Reply:
Attention weight is calculated by apply softmax for attention score.
For now NNTrainer do not support inplace operation for softmax.
I will check if inplace operation for softmax is possible(I guess it is) and if it is I will make a another pr.